onnxruntime/core/providers/cpu/tensor/slice.cc

```diff
@@ -235,7 +235,7 @@ void SliceBase::FillVectorsFromInput(const OpKernelContext* context,
   ORT_ENFORCE(nullptr != ends_tensor && ends_tensor->Shape().NumDimensions() == 1, "Ends must be a 1-D array");
   ORT_ENFORCE(start_tensor->Shape() == ends_tensor->Shape(), "Starts and ends shape mismatch");
   ORT_ENFORCE(nullptr == axes_tensor || start_tensor->Shape() == axes_tensor->Shape(), "Starts and axes shape mismatch");
-  ORT_ENFORCE(nullptr == steps_tensor || steps_tensor->Shape() == axes_tensor->Shape(), "Steps and axes shape mismatch");
+  ORT_ENFORCE(nullptr == steps_tensor || start_tensor->Shape() == steps_tensor->Shape(), "Starts and steps shape mismatch");
```

Comment:
compare with required 'starts' input

Reply:
Given the operator description says starts, ends and steps match entries in axes, would it be clearer to compare each against axes instead and for the error messages to mention axes?

You can maybe cut it down to:
- check axes is 1D if provided. if no provided create shape for default axes value
- check starts shape matches axes shape
- check ends shape matches axes shape
- check steps shape matches axes shape

---
In reply to: [279929647](https://github.com/Microsoft/onnxruntime/pull/945#discussion_r279929647) [](ancestors = 279929647)

Reply:
Since axes is optional, we compare against the (required) starts. The only difference between what you suggest and what is being done is the entity being compared against. In this way, we avoid the housekeeping like creating shape for default axes values if not present and the error messages look clear to me.

For example, hypothetically if starts = [1, 2] and ends = [3] and axes (after manual construction) become [0, 1] - I'd rather point out that starts and ends don't have corresponding values (i.e.) Starts and ends shape mismatch, rather than saying ends shape don't match axes shape when the user didn't even specify axes to begin with.